### ISSUE TITLE
`x^(3/4)` fails when x==0 but not when 0 \in x

### ISSUE THREAD

**Issue by fph** [JuliaDiff/TaylorSeries.jl#311]
This behavior seems quite counterintuitive to me:
```julia
julia> using TaylorSeries, IntervalArithmetic

julia> x = Taylor1([Interval(0),1]); x^(3/4)
ERROR: DomainError with  [1, 1] t + 𝒪(t²):
The 0th order Taylor1 coefficient must be non-zero
to raise the Taylor1 polynomial to a non-integer exponent.
Stacktrace:
 [1] pow!
   @ ~/.julia/packages/TaylorSeries/YHybm/src/power.jl:179 [inlined]
 [2] ^(a::Taylor1{Interval{Float64}}, r::Float64)
   @ TaylorSeries ~/.julia/packages/TaylorSeries/YHybm/src/power.jl:109
 [3] top-level scope
   @ REPL[20]:1

julia> x = Taylor1([Interval(0,1e-15),1]); x^(3/4)
 [0, 5.62342e-12] + [0, ∞] t + 𝒪(t²)

julia> x = Taylor1([Interval(-1e-15,1e-15),1]); x^(3/4)
 [0, 5.62342e-12] + [-∞, ∞] t + 𝒪(t²)
```
The computation fails if x is equal to 0, but it succeeds (with an arguably correct result)  in the more general case in which x is an interval that contains 0. 

Actually returning a series with an infinite derivative seems more useful (at least in my use case), so I would suggest to simply remove the check for 0 and let TaylorSeries return infinite derivatives also in the first case.

Regarding the last example, from what I understand the negative part of the interval is ignored by design; however I found it unusual that the method returns a larger enclosure for the derivative; I would have expected that computation to be equivalent to first clamping the interval to 0.

**Comment by lbenet** [comment]
Hi, and thanks for opening this issue! 

TLDR; I think you are right, there is some counterintuitive results, though maybe not all: the first one of your examples is actually ok, but the third indeed needs consistent shrinking of the interval of the 0-th order coefficient of x for the computation of higher order terms.

Let me explain first the way the computations are obtained: the k-th coefficient of `c=a^r` (`a::Taylor1`) is obtained using the `pow!` (internal) function, see https://juliadiff.org/TaylorSeries.jl/latest/api/#TaylorSeries.pow! , which includes the recursion relation used. Clearly, the appearance of `a_0` (the 0th-order coefficient of `a`) in the denominator is the source of issues when it is *identical* to zero. To avoid that, which otherwise means obtaining `Inf`s or `NaN`s for the remaining coefficients, we check if that coefficient is zero, and throw an error if that is the case. This, I think, is consistent (*intuitive*) for Taylor coefficients that are "usual numbers" (real or complex); having coefficients that are intervals is a bit more subtle.

Now I shall argue that the very first case you have above, `x = Taylor1([Interval(0),1]); x^(3/4)`, is "arguably correct". The 0-th coefficient of `x` is zero, and thus the0-th coefficient of the result has to be zero; using the formula in the docs, the 1st order reduces to compute `Interval(0)/Interval(0)`, which results in the empty set! Since any arithmetic operation with empty sets returns an empty set, I think to throw an error makes sense, as it is done, though I agree that in this case it is not the best error message...

The second example seems ok, as you point out. Once said that, the fact that the 1st-order coefficient is `[0, ∞]` has to be considered with some care: Intervals, in the "set-based" definition, are "closed and connected", which imposes that the lower and upper bounds are both *real* numbers (in the mathematical sense); so having `Inf` as one of the bounds is a source of many discussions (is it infinity, in a mathematical sense, or a very very large real number  represented as "Inf" as imposed by the closeness of the intervals?).

Regarding the third example, I think your point is correct: the 0th order coefficient of the result is obtained *after* the intersection (of the 0-th order coefficient of `x`) with a "natural domain" (of the `pow` function), which is `Interval(0, Inf)`, but higher other coefficients (of the result) do not use the shrunk `a_0` coefficient. This certainly has to be corrected; maybe for other functions need this too, such as `log` (which also involves the intersection with some natural domain). I hope to find some time soon to make such changes.

**Comment by lbenet** [comment]
I'm glad that the answer is helpful! Indeed, the issue here with `^` is pretty similar to that in #302; thanks for including those issues here.

Let me guide you a little bit on how to hard-code `mtsqrt`, so it behaves as you like; the following follows [this comment](https://github.com/JuliaDiff/TaylorSeries.jl/issues/302#issuecomment-1186465315):
```julia
function mysqrt(a::Taylor1)

    aux = zero(sqrt( constant_term(a) ))

    c_order = a.order
    c = Taylor1( aux, c_order )
    aa = one(aux) * a
    for k = 0:a.order
        TaylorSeries.sqrt!(c, aa, k, 0)
    end

    return c
end
```
With this function, you get:
```julia
julia> t = Taylor1(6)
1.0 t + 𝒪(t⁷)

julia> mysqrt(t)
Inf t - Inf t² + Inf t³ - Inf t⁴ + Inf t⁵ - Inf t⁶ + 𝒪(t⁷)

julia> ti = Taylor1([Interval(0), Interval(1)], 6)
 [1, 1] t + 𝒪(t⁷)

julia> mysqrt(ti)
 ∅ t + ∅ t² + ∅ t³ + ∅ t⁴ + ∅ t⁵ + ∅ t⁶ + 𝒪(t⁷)

julia> ti1 = Taylor1([Interval(0, 1.e-15), Interval(1)], 6)
 [0, 1.00001e-15] + [1, 1] t + 𝒪(t⁷)

julia> mysqrt(ti1)
 [0, 3.16228e-08] + [1.58113e+07, ∞] t + [-∞, -3.95284e+21] t² + [1.97642e+36, ∞] t³ + [-∞, -1.23526e+51] t⁴ + [8.64685e+65, ∞] t⁵ + [-∞, -6.48513e+80] t⁶ + 𝒪(t⁷)

julia> ti2 = Taylor1([Interval(-1.e15, 1.e-15), Interval(1)], 6)
 [-1e+15, 1.00001e-15] + [1, 1] t + 𝒪(t⁷)

julia> mysqrt(ti2)
 [0, 3.16228e-08] + [1.58113e+07, ∞] t + [-∞, -3.95284e+21] t² + [1.97642e+36, ∞] t³ + [-∞, -1.23526e+51] t⁴ + [8.64685e+65, ∞] t⁵ + [-∞, -6.48513e+80] t⁶ + 𝒪(t⁷)
```
Note that the current version in master of `sqrt` essentially behaves as `mysqrt`; the difference lies in throwing the error.

You can do something similar for `mypow`, though in that case you have to make sure that the domain restriction is consistently implemented (so you use the intersection of the constant term of the series with `Interval(0, Inf)`), and you have to to adapt `TaylorSeries.pow!` too. The reason you have to modify `TaylorSeries.pow!` too, and not `TaylorSeries.sqrt!`, resides in the recursion formulae: for [`sqrt!`](https://juliadiff.org/TaylorSeries.jl/dev/api/#TaylorSeries.sqrt!) you use $c_0$ while for [`pow!`](https://juliadiff.org/TaylorSeries.jl/dev/api/#TaylorSeries.pow!) you use $a_0$.

**Comment by lbenet** [comment]
I've open #312 as a possible solution for this issue, and extending it to other functions that need some domain restrictions; that PR still needs some tests...

Related to the last remark abovem on the changes on `sqrt!` and `pow!`, the following illustrates some of the subtleties that arise:
```julia
julia> sqrt(sqrt(ti1))
 [0, 0.000177828] + [4.44569e+10, ∞] t + [-∞, -1.66713e+25] t² + [9.72496e+39, ∞] t³ + [-∞, -6.68591e+54] t⁴ + [5.01443e+69, ∞] t⁵ + [-∞, -3.96976e+84] t⁶ + 𝒪(t⁷)

julia> ti1^0.25
 [0, 0.000177828] + [0, ∞] t + [-∞, 0] t² + [0, ∞] t³ + [-∞, 0] t⁴ + [0, ∞] t⁵ + [-∞, 0] t⁶ + 𝒪(t⁷)

julia> sqrt(sqrt(ti2))
 [0, 0.000177828] + [4.44569e+10, ∞] t + [-∞, -1.66713e+25] t² + [9.72496e+39, ∞] t³ + [-∞, -6.68591e+54] t⁴ + [5.01443e+69, ∞] t⁵ + [-∞, -3.96976e+84] t⁶ + 𝒪(t⁷)

julia> ti2^0.25
 [0, 0.000177828] + [0, ∞] t + [-∞, 0] t² + [0, ∞] t³ + [-∞, 0] t⁴ + [0, ∞] t⁵ + [-∞, 0] t⁶ + 𝒪(t⁷)
```
Note simply that `sqrt(sqrt(x))` is "tighter" than `x^0.25`.

**Comment by lbenet** [comment]
Closed by #312
